pinot-core/src/main/java/org/apache/pinot/core/operator/filter/predicate/PredicateUtils.java

```diff
@@ -139,11 +144,22 @@ public static IntSet getDictIdSet(BaseInPredicate inPredicate, Dictionary dictio
         }
         break;
       case STRING:
-        for (String value : values) {
-          int dictId = dictionary.indexOf(value);
-          if (dictId >= 0) {
-            dictIdSet.add(dictId);
+        if (queryContext == null || values.size() < Integer.parseInt(queryContext.getQueryOptions()
```

Comment:
(minor) Suggest changing to `<=` for consistency, we usually treat threshold as exclusive